src/message_sender.pyx

```diff
@@ -99,13 +105,21 @@ cdef void on_message_send_complete(void* context, c_message_sender.MESSAGE_SEND_
         send_data = c_amqpvalue.amqpvalue_clone(delivery_state)
         wrapped = copy.deepcopy(value_factory(send_data).value)
     if context != NULL:
+        context_pyobj = <PyObject*>context
+        if context_pyobj.ob_refcnt == 0: # context is being garbage collected, skip the callback
+            _logger.warning("Can't call _on_message_sent during garbage collection")
+            return
         context_obj = <object>context
         if hasattr(context_obj, "_on_message_sent"):
             context_obj._on_message_sent(context_obj, send_result, delivery_state=wrapped)
 
 
 cdef void on_message_sender_state_changed(void* context, c_message_sender.MESSAGE_SENDER_STATE_TAG new_state, c_message_sender.MESSAGE_SENDER_STATE_TAG previous_state):
     if context != NULL:
+        context_pyobj = <PyObject*>context
+        if context_pyobj.ob_refcnt == 0: # context is being garbage collected, skip the callback
+            _logger.warning("Can't call _state_changed during garbage collection")
```

Comment:
Wont this be part of a normal shutdown? I'm not sure it should be logged as warning...

Reply:
Nope, only if you didn't use the context manager nor the close.

Still, as discussed offline will update a little the message